### PR TITLE
4703-dispensing

### LINKED ISSUE #4703: Use different default sell prices for patient dispensing vs outbound shipments
## Is your feature request related to a problem? Please describe 👀

A requested use case is for one default sell price to be used for when shipping stock to downstream facilities and a different default sell price to be used when dispensing to patients. 

Some Stores may do both - e.g. a hospital sending stock to smaller pharmacies and also dispensing directly to patients. 

## Describe the solution you'd like 🎁

Here's the pricing workflow:

image

## Configuration
( All are items are changes to make in mSupply )
* Create a master list for all items, check the `is_default_price_list` option
* This `is_default_price_list` applies to all stores, and doesn't have to be assigned to them...
* Define prices for each item 
* Create a master list for all discounted items, set the `discount` percentage value
* This discount master list need to be assigned to any stores that should receive the discount rate.
 
## Changes required

* update master lists
  * add `is_default_price_list` to `master_list`
  * add `discount` to `master_list` 
  * and update sync and translation layers
  * add `price` to `master_list_line` and update sync & translation layers

* When stock is issued apply the `is_default_price_list` prices to the invoice line (as appropriate)
* Apply discount to outgoing invoices (apart from dispensing)

**Stocktake**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] If new stock line is introduced by stocktake, it's sell price should be set to price from `is_default_price_list`
- [ ] Stocklines are introduced from either Add Item, or Edit Item (if the stock take is created from a master list) (FRONTEND)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**View Stock**
_No changes_

...OUT OF SCOPE - POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell Price should be prepopulated based on `is_default_price_list` once a Item is selected in the New Stock Modal
![image](https://github.com/user-attachments/assets/a5ba6642-cca7-4699-a647-5799c6168471)
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Outbound Shipment**
- [x] Sell price should looked up from the `is_default_price_list`
- [x] However, a __discount__ should be applied based on the Customer's(Name) discount master list
   - [x] The global discount should apply from the masterlist to all items that are in that discount master list
   - [x] If an item is not in the discount master list no discount should be applied
- [x] Add Item (Frontend) => Modal runs graphql query to get Discount & apply to `invoice_line`)
- [x] Actions-> Allocate Lines (Backend: `allocatedOutboundShipmentUnallocatedLines`)  => Backend to look up default price list and apply price + any discount automatically
- [x] Add from Master List (Backend:`addToOutboundShipmentFromMasterList` ) => _No Changes required as pricing will be applied at allocation time_

**Inbound Shipment**
_No changes_

...OUT OF SCOPE -  POSSIBLE FUTURE REQUIREMENTS...
```
- [ ] Sell price should looked up from the `is_default_price_list`
- [ ] Discount will be applied based on the store name's discount list whether it comes via a transfer or otherwise
- [ ] Add from Master List (`addToInboundShipmentFromMasterList`) => look up `is_default_price_list` price for that item and apply it to `sell_price` 
- [ ] Cost price should come from sell price on inbound shipment
- [ ] Add Item  Frontend) => Modal runs graphql query to get`is_default_price_list` and applies it to `Sell Price`
- [ ] Cost prices are assumed to be discounted if discount would apply to this store
- [ ] Sell Prices are not editable by the user 
```

**Internal Orders**
_No changes required_

**Requsitions, R&R**
_No changes required_

**Returns**
_No changes required_

**Transfer aka in Processors code**
_No changes required_

**Dispensing**
- [x] Sell price should looked up from the `is_default_price_list` (Stockline Price is ignored)
- [x] No discounts are applied during dispensing

### Additional context 💌

Related to https://github.com/msupply-foundation/msupply/issues/15108

### Moneyworks Jobcode 🧰

`COMSCUST`

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #4703

# 👩🏻‍💻 What does this PR do?

Puts an exception in for dispensing to ignore discounts

## 💌 Any notes for the reviewer?
 
Will do one more PR with clean ups from previous PR.

# 🧪 Testing

- [ ] Setup mSupply to a branch including #15213-default-price-list-and-discount-pricing
- [ ] Setup a discount list and default price list in mSupply
- [ ] Check default price list is used when dispensing to patients, no discount should be applied

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole

### CODE REVIEW

File: server/service/src/pricing/item_price.rs [hunk 1]
```diff
@@ -50,19 +50,36 @@ pub fn get_pricing_for_item(
...
-        .query_by_filter(
-            MasterListFilter::new()
-                .is_discount_list(true)
-                .item_id(EqualFilter::equal_to(&input.item_id)),
-        )?
-        .pop();
+    // 2. Check if we have a name, and that name is not a patient
+    let is_patient = match &input.customer_name_id {
+        Some(customer_name_id) => {
+            let num_patients = PatientRepository::new(&ctx.connection).count(
```
Comment (andreievg) [inline server/service/src/pricing/item_price.rs:56]: There is `query_one` on PatientRepository and there is also `is_some()` on Option

Comment (jmbrunskill) [reply server/service/src/pricing/item_price.rs:56]: I thought this was arguably more efficient for the database, happy either way though...

**Review by andreievg — APPROVED** [review]
https://github.com/msupply-foundation/open-msupply/blob/747640f7932e0f7de670c01f628faeb116f60ff6/server/service/src/pricing/item_price.rs#L11-L12

It's used now

Not requesting this change, at it would require a few other changes futher on, but wanted to emphasise a point i've been feeling, it's about verboseness of rust being intermidating factor and I think a hurdle for rust beginers (which I think we need to be thinking about going forwards). I am not sure if below suggestion really deals with it, but it's food for though for futured dev, i think it's less code and easier to reason about (basically not having optionals but just defaulting to 0, adding master list filter to master list line and adding query_by_one to repositories. Below suggestion won't compile but i wanted to show an example of how it could look:

https://github.com/msupply-foundation/open-msupply/compare/4703-dispensing...4703-dispensing(Example) -> would suggest a refactor issue for this

Thanks for this PR though, I am happy to approve

**Comment by jmbrunskill** [comment]
> It's used now

Nice good catch.

I see your point with not making thing optional. I'm a fan of minimising how often we do Optionals as well.
For some reason it feels like we should distinghish between a 0.0% discount and no discount some how but I think I agree with you that it's probably better to make non-optional.

**Pull request merged**